google/cloud/storage/internal/object_requests.cc

```diff
@@ -159,6 +159,79 @@ std::ostream& operator<<(std::ostream& os, UpdateObjectRequest const& r) {
   return os << "}";
 }
 
+PatchObjectRequest::PatchObjectRequest(std::string bucket_name,
+                                       std::string object_name,
+                                       ObjectMetadata const& original,
+                                       ObjectMetadata const& updated)
+    : GenericObjectRequest(std::move(bucket_name), std::move(object_name)) {
+  // Compare each writeable field to build the patch.
+  ObjectMetadataPatchBuilder builder;
+
+  if (original.acl() != updated.acl()) {
+    builder.SetAcl(updated.acl());
+  }
+  if (original.cache_control() != updated.cache_control()) {
+    builder.SetCacheControl(updated.cache_control());
+  }
+  if (original.content_disposition() != updated.content_disposition()) {
+    builder.SetContentDisposition(updated.content_disposition());
+  }
+  if (original.content_encoding() != updated.content_encoding()) {
+    builder.SetContentEncoding(updated.content_encoding());
+  }
+  if (original.content_language() != updated.content_language()) {
+    builder.SetContentLanguage(updated.content_language());
+  }
+  if (original.content_type() != updated.content_type()) {
+    builder.SetContentType(updated.content_type());
+  }
+
+  if (original.metadata() != updated.metadata()) {
+    if (updated.metadata().empty()) {
+      builder.ResetMetadata();
+    } else {
+      std::map<std::string, std::string> difference;
+      // Find the keys in the original map that are not in the new map:
+      std::set_difference(original.metadata().begin(),
+                          original.metadata().end(), updated.metadata().begin(),
+                          updated.metadata().end(),
+                          std::inserter(difference, difference.end()),
+                          // We want to compare just keys and ignore values, the
+                          // map class provides such a function, so use it:
+                          original.metadata().value_comp());
```

Comment:
I spent a bit of time hunting this down -- `value_comp()` works here because std::map's iterator goes in ascending order of keys (smallest is at `begin()`, largest is at `rbegin()`).
https://stackoverflow.com/questions/7648756/is-the-order-of-iterating-through-stdmap-known-and-guaranteed-by-the-standard

If that's correct, could you leave a comment mentioning it?  I've asked a couple other people, and they weren't aware of this either :P